### PR TITLE
Welcome Tour Next: Base functionality - step repositioning & spotlight

### LINKED ISSUE #56353: Editor Step-by-step guides: Tech foundation
**Update:** before trying anything new in this direction, would be good to go through the previous spike discussed in pbxlJb-eL-p2 and implemented using Shepherd in https://github.com/Automattic/wp-calypso/pull/43365

___

- copy to an a8c package just the utils / components (most of them are in [config-elements](https://github.com/Automattic/wp-calypso/tree/trunk/client/layout/guided-tours/config-elements)) -  that are mandatory for a step-by-step guide to work inside the editor.
- import them in ETK to create a proof of concept guided tour:
  * it may replace Welcome Tour showing some dummy content or it may move it around entirely
  * have at least 2 steps displayed on 2 different UI elements
  * activate it based on some feature flag

### LINKED ISSUE #57056: Welcome Tour NEXT: Move core functionality behind a feature flag
### Details

As of #56867 the Welcome Tour applies step repositioning to move the steps near reference elements on the page  (or fallback to the default position - bottom left). It also adds a semi-transparent overlay to block interaction with the editor.

Let's move this functionality behind a feature flag and ship it. Something like `?tour_next` might work.

### Checklist

- [ ] Using the feature flag and opening the welcome guide should show the step-tour instead of the current/static one.
- [ ] Not using the feature flag should open the current/static tour.

### Related

_No response_

### PR DESCRIPTION
**Pull request by chriskmnds**
A feature PR that introduces step repositioning and spotlight effect to the welcome tour. Introduces the new functionality behind a URL param for testing.

#### Changes proposed in this Pull Request

Sets the ground for moving/positioning the tour near the element the various steps refer to - so going from a static tour to a more step-by-step guided tour. Main concepts being introduced:

1. Popper.js to move the tour near reference elements on the page.
2. A dark overlay that blocks mouse events from reaching the editor. Focus is trapped inside the tour.
3. A spotlight effect to highlight reference elements behind the overlay. 
4. Hides the functionality behind a search param (`welcome-tour-next`).

#### Media

Kapture 2021-11-08 at 17 00 04

#### Testing instructions

1. Sandbox a site, checkout branch, and `yarn dev --sync` from `apps/editing-toolkit` to sync sandbox.
2. Go to the editor, append `&welcome-tour-next` to the URL, reload. 
3. Open the sidebar and from there click on "Welcome Guide".
3. Tour should be focused and keyboard navigation enabled.
4. Focus should be on the "Try it out!" button.
5. Stepping through the tour should move steps closer to the element they refer to.
6. Reference elements should be highlighted, as in the demo.
7. Navigating via Tab key should trap the focus within the tour if the tour is expanded (not minified).
8. Minifying the tour should show the editor.
9. Removing `&welcome-tour-next` from the URL and trying again should open the previous tour (no step repositioning, no dark overlay, no spotlight/highlighting of reference elements).

#### TODO

This is a foundation branch. Current and new issues being added to a GH project: https://github.com/Automattic/wp-calypso/milestone/365

Closes #56353 #57056

### CODE REVIEW

**Comment by chriskmnds** [comment]
@poligilad-auto @razvanpapadopol - once this is merged, we should have the new functionality (steps, overlay, spotlight) behind a URL param `&welcome-tour-next`. We'll be reviewing with @escapemanuele and aim to ship soon so we can have it live and unblock further work on it.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js [hunk 1]
```diff
@@ -112,63 +116,122 @@ function WelcomeTourFrame() {
 		setJustMaximized( true );
 		recordTracksEvent( 'calypso_editor_wpcom_tour_maximize', {
 			is_gutenboarding: isGutenboarding,
-			slide_number: currentCardIndex + 1,
+			slide_number: currentStepIndex + 1,
 		} );
 	};
 
+	const isWelcomeTourNext = () => {
+		new URLSearchParams( document.location.search ).has( 'welcome-tour-next' );
```
Comment (escapemanuele) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js]: I think that we should add a `return` here 😄

**Review by escapemanuele — COMMENTED** [review]
If we add `return` to `isWelcomeTourNext` it should work 👍

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/components/spotlight/index.tsx [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import { usePopper } from 'react-popper';
+import type { Rect, Placement } from '@popperjs/core';
+
+interface Props {
+	referenceElementSelector: string;
+}
+
+const Spotlight: React.FunctionComponent< Props > = ( { referenceElementSelector } ) => {
+	const popperElementRef = useRef( null );
+	const referenceElement = document.querySelector( referenceElementSelector );
```
Comment (escapemanuele) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/components/spotlight/index.tsx:15]: Could it be the same if we pass directly `referenceElement` from `tour-launch.js`, without using`document.querySelector` here? Or you did this way for performance reasons (passing Element as a prop)? :)

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/hooks/use-focus-trap.ts [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+			if ( handled ) {
+				event.preventDefault();
+				event.stopPropagation();
+			}
+		},
+		[ firstFocusableElement, lastFocusableElement ]
+	);
+
+	useEffect( () => {
+		const focuableElementSelectors = [
```
Comment (escapemanuele) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/hooks/use-focus-trap.ts]: `focusableElementSelectors`, just a typo 👍

**Review by escapemanuele — COMMENTED** [review]
Just two minor things, great job in this PR 💪 
I think that it can be merged after the `return` fix

**Review by escapemanuele — APPROVED** [review]
Looks good to me 👍

**Comment by noahtallen** [comment]
Just a heads up, this added a package-lock.json file under apps/editing-toolkit! Will delete it, but you can run `yarn` from the root of wp-calypso and it should install anything needed for ETK

**Pull request merged**